app/lib/package/api_export/exported_api.dart

```diff
@@ -0,0 +1,327 @@
+  }
+}
+
+/// Interface for an exported file.
+sealed class ExportedObject {
+  final ExportedApi _owner;
+  final String _objectName;
+  ExportedObject._(this._owner, this._objectName);
+
+  /// Delete this file.
+  Future<void> delete() async {
+    await Future.wait(_owner._prefixes.map((prefix) async {
+      await _owner._pool.withResource(() async {
+        await _owner._bucket.delete(prefix + _objectName);
+      });
+    }));
+  }
+}
+
+/// Interface for an exported JSON file.
+///
+/// This will write JSON as gzipped UTF-8, adding headers for
+///  * `Content-Type`,
+///  * `Content-Encoding`, and,
+///  * `Cache-Control`.
+final class ExportedJsonFile extends ExportedObject {
+  static final _jsonGzip = json.fuse(utf8).fuse(gzip);
+  final Duration _maxAge;
+
+  ExportedJsonFile._(
+    ExportedApi _owner,
+    String _objectName,
+    this._maxAge,
+  ) : super._(_owner, _objectName);
+
+  late final _metadata = ObjectMetadata(
+    contentType: 'application/json; charset="utf-8"',
+    contentEncoding: 'gzip',
+    cacheControl: 'public, max-age=${_maxAge.inSeconds}',
+  );
+
+  /// Write [data] as gzipped JSON in UTF-8 format.
+  Future<void> write(Map<String, Object?> data) async {
+    final gzipped = _jsonGzip.encode(data);
+    await Future.wait(_owner._prefixes.map((prefix) async {
+      await _owner._pool.withResource(() async {
+        await _owner._bucket.writeBytes(
+          prefix + _objectName,
+          gzipped,
+          metadata: _metadata,
+        );
+      });
+    }));
+  }
+}
+
+/// Interface for an exported binary file.
+///
+/// This will write a binary blob as is, adding headers for
+///  * `Content-Type`,
+///  * `Content-Disposition`, and,
+///  * `Cache-Control`.
+final class ExportedBlob extends ExportedObject {
+  final String _contentType;
+  final Duration _maxAge;
+  final String _filename;
+
+  ExportedBlob._(
+    ExportedApi _owner,
+    String _objectName,
+    this._filename,
+    this._contentType,
+    this._maxAge,
+  ) : super._(_owner, _objectName);
+
+  late final _metadata = ObjectMetadata(
+    contentType: _contentType,
+    cacheControl: 'public, max-age=${_maxAge.inSeconds}',
+    contentDisposition: 'attachment; filename="$_filename"',
+  );
+
+  /// Write binary blob to this file.
+  Future<void> write(List<int> data) async {
+    await Future.wait(_owner._prefixes.map((prefix) async {
+      await _owner._pool.withResource(() async {
+        await _owner._bucket.writeBytes(
+          prefix + _objectName,
+          data,
+          metadata: _metadata,
+        );
+      });
+    }));
+  }
+
+  /// Copy binary blob from [absoluteObjectName] to this file.
+  ///
+  /// Notice that [absoluteObjectName] must be an a GCS URI including `gs://`.
+  /// This means that it must include bucket name.
+  /// Such URIs can be created with [Bucket.absoluteObjectName].
+  Future<void> copyFrom(String absoluteObjectName) async {
+    await Future.wait(_owner._prefixes.map((prefix) async {
```

Comment:
Is it possible that we have a race condition here which may update the second object earlier? (Is that bad?) -- actually this applies for most of the similar Future.wait calls in this file...